### PR TITLE
can select car number now

### LINKED ISSUE #3556: [Finance] - Finance Dashboard Select Car Number
### Description

Users should be able to select a car number to view financial data about a specific car

### Acceptance Criteria

.

### Proposed Solution

.

### Mocks

_No response_

### PR DESCRIPTION
**Pull request by chpy04**
## Changes

There is a dropdown to select a car as well as the dates. For everything to do with work packages, the car number will cause only work packages on that car to be returned. In the case of other reason rrs, it will use the start and end dates provided. If none are provided, it will use the creation date of the selected car as the start date, and the creation date of the car after the selected car as the end date.

## Notes

_Any other notes go here_

## Test Cases

- Case A
- Edge case
- ...

## Screenshots

_If you made UI changes you must post a screenshot of the whole page for each change in 1) a normal sized window and 2) the smallest possible window_

_If you did any manual testing (e.g., with Postman), put screenshots of the http request and before and after of the db_

_If none of this applies, you can delete this section_

## To Do

_Any remaining things that need to get done_

- [ ] item 1
- [ ] ...

## Checklist

It can be helpful to check the `Checks` and `Files changed` tabs.
Please review the [contributor guide](https://nerdocs.atlassian.net/wiki/spaces/NER/pages/8060929/Software+Contributor+Guide) and reach out to your Tech Lead if anything is unclear.
Please request reviewers and ping on slack only after you've gone through this whole checklist.

- [ ] All commits are tagged with the ticket number
- [ ] No linting errors / newline at end of file warnings
- [ ] All code follows repository-configured prettier formatting
- [ ] No merge conflicts
- [ ] All checks passing
- [ ] Screenshots of UI changes (see Screenshots section)
- [ ] Remove any non-applicable sections of this template
- [ ] Assign the PR to yourself
- [ ] No `yarn.lock` changes (unless dependencies have changed)
- [ ] Request reviewers & ping on Slack
- [ ] PR is linked to the ticket (fill in the closes line below)

Closes #3556

### CODE REVIEW

File: src/backend/src/controllers/finance.controllers.ts [hunk 1]
```diff
@@ -264,11 +280,18 @@ export default class FinanceController {
 
   static async getAllSpendingBarData(req: Request, res: Response, next: NextFunction) {
     try {
-      const { startDate, endDate } = req.query;
+      const { startDate, endDate, carNumber } = req.query;
+      console.log(carNumber);
```
Comment (walker-sean) [inline src/backend/src/controllers/finance.controllers.ts]: Left this in

File: src/backend/src/services/finance.services.ts [hunk 2]
```diff
@@ -407,72 +398,770 @@ export default class FinanceServices {
...
+        },
+        ...getReimbursementRequestWhereInput(startDate, endDate)
+      },
+      select: {
+        reimbursementStatuses: true,
+        totalCost: true
+      }
+    });
+
+    let pendingFinance = 0;
```
Comment (walker-sean) [inline src/backend/src/services/finance.services.ts]: This would be an awesome place to use a map ( / object) with keys as statuses and then we wouldn't have to use a switch statement at all

Comment (walker-sean) [reply src/backend/src/services/finance.services.ts]: And then the values could be extracted using object destructuring

File: src/backend/src/services/finance.services.ts [hunk 3]
```diff
@@ -407,72 +398,770 @@ export default class FinanceServices {
...
+      },
+      select: {
+        reimbursementStatuses: true,
+        totalCost: true
+      }
+    });
+
+    const totalBudget = team.projects.reduce((acc, curr) => acc + curr.budget, 0);
+
+    let pendingFinance = 0;
```
Comment (walker-sean) [inline src/backend/src/services/finance.services.ts]: Same comment as before -- since this is repeated code it would be nice to make this a helper

File: src/backend/src/services/finance.services.ts [hunk 4]
```diff
@@ -407,72 +398,770 @@ export default class FinanceServices {
...
+      budgetReimbursementRequests.reduce((reqAcc, rr) => {
+        return reqAcc + rr.totalCost;
+      }, 0) / 100;
+
+    let allPendingFinance = 0;
+    let allPendingLeadership = 0;
+    let allSubmittedToSabo = 0;
+    let allReimbursed = 0;
+
+    allReimbursementRequests.forEach((req) => {
```
Comment (walker-sean) [inline src/backend/src/services/finance.services.ts]: Same

File: src/backend/src/services/finance.services.ts [hunk 5]
```diff
@@ -407,72 +398,770 @@ export default class FinanceServices {
...
+    const allTotalBalance = allReimbursementRequests.reduce((acc, curr) => acc + curr.totalCost, 0) / 100;
+
+    const allAvailable = allTotalBudget - allTotalBalance;
+
+    let cashPendingFinance = 0;
+    let cashPendingLeadership = 0;
+    let cashSubmittedToSabo = 0;
+    let cashReimbursed = 0;
+
+    cashReimbursementRequests.forEach((req) => {
```
Comment (walker-sean) [inline src/backend/src/services/finance.services.ts]: Same

File: src/backend/src/services/finance.services.ts [hunk 6]
```diff
@@ -407,72 +398,770 @@ export default class FinanceServices {
...
+    const cashTotalBalance = cashReimbursementRequests.reduce((acc, curr) => acc + curr.totalCost, 0) / 100;
+
+    const cashAvailable = cashTotalBudget - cashTotalBalance;
+
+    let budgetPendingFinance = 0;
+    let budgetPendingLeadership = 0;
+    let budgetSubmittedToSabo = 0;
+    let budgetReimbursed = 0;
+
+    budgetReimbursementRequests.forEach((req) => {
```
Comment (walker-sean) [inline src/backend/src/services/finance.services.ts]: Same

File: src/backend/src/services/finance.services.ts [hunk 7]
```diff
@@ -407,72 +398,770 @@ export default class FinanceServices {
...
+    if (!category) throw new NotFoundException('Reimbursement Product Other Reason', otherReasonId);
+
+    const totalBudget = category.budget;
+
+    let pendingFinance = 0;
+    let pendingLeadership = 0;
+    let submittedToSabo = 0;
+    let reimbursed = 0;
+
+    reimbursementRequests.forEach((req) => {
```
Comment (walker-sean) [inline src/backend/src/services/finance.services.ts:933]: Same

File: src/backend/src/services/finance.services.ts [hunk 8]
```diff
@@ -407,72 +398,770 @@ export default class FinanceServices {
...
     startDate?: Date,
-    endDate?: Date
+    endDate?: Date,
+    carNumber?: number
   ): Promise<ReimbursementRequestData> {
-    return await getReimbursementRequestCategoryData(otherReasonId, organization.organizationId, startDate, endDate);
+    const { organizationId } = organization;
+
+    //if a car is specified but not the date ranges, use the car to determine the date ranges
+    if (carNumber !== undefined && (!startDate || !endDate)) {
```
Comment (walker-sean) [inline src/backend/src/services/finance.services.ts]: We probably want to be able to filter by car and date together (intersection)

**Review by walker-sean — CHANGES_REQUESTED** [review]

**Review by walker-sean — APPROVED** [review]
